bpo-24139: Add support for SQLite extended result codes

erlend-aasland opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-24139](https://bugs.python.org/issue24139) -->
https://bugs.python.org/issue24139
<!-- /issue-number -->

Commit:
Add extended result codes as module constants

Commit:
Use extended result code to add exception attributes

Commit:
Add ifdefs for extended error codes

Commit:
Add test for extended error codes

Commit:
Add NEWS item

Commit:
Fix windows tests

Commit:
Fix ifdefs for extended result codes

Commit:
Fix tests again

Commit:
Add extended error codes from upcoming SQLite 3.37.0

Commit:
Squeeze a couple of lines in test_module_constants()

Commit:
Merge branch 'main' into sqlite-extended-error-codes

pablogsal commented on Modules/_sqlite/module.c:
```diff
@@ -320,6 +320,115 @@ static const struct {
     DECLARE_ERROR_CODE(SQLITE_NOTICE),
     DECLARE_ERROR_CODE(SQLITE_WARNING),
+#endif
+    // Extended result code list
```
Can you add here some link to the docs so we can check in the future more codes and the version constraints?

erlend-aasland replied:
Will do!

erlend-aasland replied:
BTW, the version constraints were a bit tricky; many of them were not mentioned in the SQLite changelog, so I used a simple shell script for grepping through all the SQLite versions, finding out when the various macros appeared.

pablogsal replied:
> so I used a simple shell script for grepping through all the SQLite versions

Then isn't it better to use  preprocessor checks to see if they are defined?

erlend-aasland replied:
That's also an option. I thought about that, but I though this was a little bit more elegant, IMO. Let me know if you want me to change it.

pablogsal replied:
I think modifying/adding a macro that adds the preprocessor checks + the addition may be a bit more future-proof.

pablogsal replied:
My main worry is how this can be extended in the future and what can happen if we do it incorrectly. With the checks, there is no possibility of getting if wrong :)

erlend-aasland replied:
True that! I'll fix it tomorrow :)

erlend-aasland replied:
_But_, if we keep the version checks as they are, it's easier to declutter the code when we up the SQLite version requirements.

pablogsal replied:
That's true. At this point, I trust your judgement. Whatever you think is best is going to be good enough :)

Just make sure to add enough comments so the next person knows how to add more stuff in the future

erlend-aasland replied:
All right; I'll think about it thrice before deciding :)

erlend-aasland replied:
I'm keeping the version checks for future decluttering. I added some comments, but I'm not satisfied with them (yet). Finding out when a new result code was added will probably involve some detective work anyway, given how the SQLite writes their changelogs.

erlend-aasland replied:
I think 26517fc8e692304c1c77a26fcbb9e6ca78989735 is ok. PTAL.

erlend-aasland replied:
Marking this as resolved. Please ping me if you disagree :)

Commit:
Address review: add link to SQLite docs

Commit:
Add links to SQLite release docs

Commit:
Fix #if's for error codes between SQLite 3.8.2 and 3.8.10

SQLITE_CONSTRAINT_ROWID, SQLITE_READONLY_DBMOVED, and SQLITE_AUTH_USER

Commit:
Add support for the SQLITE_RANGE primary result code

The SQLITE_RANGE error indices that the parameter number argument to one
of the sqlite3_bind routines or the column number in one of the
sqlite3_column routines is out of range.  This should be considered a
misuse of the SQLite API, so we translate it to InterfaceError, as
defined by PEP 249.

Commit:
Fix test_extended_error_code_on_exception version requirement

Commit:
What's New

Commit:
Use managed connect in the added test

Commit:
Merge branch 'main' into sqlite-extended-error-codes

Commit:
Improve comment

Commit:
Merge branch 'main' into sqlite-extended-error-codes

Commit:
Remove 3.37.0 (dev) error codes. 3.37 is still in development

Commit:
Remove dev (3.37.0) flags from tests as well
